### PR TITLE
Speed up function `group_tools`

### LINKED ISSUE #14748: Performance improvement to group_tools
## Summary

Optimizes the `group_tools` function in `src/bokeh/layouts.py` through two key changes:

1. **Move dataclass to module level**: Moving the `ToolEntry` dataclass from inside the function to module level eliminates the overhead of redefining the dataclass on every function call.

2. **Replace O(n²) list removal with O(n) boolean array**: The original code used `entries.remove(item)` repeatedly within nested loops, which is O(n) per removal operation. The optimization replaces this with a boolean `used` array that marks processed entries in O(1) time.

## Performance Impact

This change speeds up `gridplot` when `merge_tools=True` by reducing the time spent in `group_tools`.

## Related PR

- #14702

### PR DESCRIPTION
**Pull request by misrasaurabh1**
#### 📄 32% (0.32x) speedup for ***`group_tools` in `src/bokeh/layouts.py`***

Saurabh's comment - Speeds up gridplot when merge_tools is True

⏱️ Runtime : **`5.41 milliseconds`** **→** **`4.10 milliseconds`** (best of `5` runs)

#### 📝 Explanation and details

The optimized code achieves a **31% speedup** through three key performance improvements:

**1. Eliminated O(n²) list removal operations (biggest impact):**
The original code used `entries.remove(item)` repeatedly within nested loops, which is O(n) per removal operation. With multiple removals in nested loops, this became O(n²). The optimization replaces this with a boolean `used` array that marks processed entries in O(1) time, reducing the overall complexity from O(n²) to O(n).

**2. Replaced @dataclass with NamedTuple:**
Changed `ToolEntry` from a `@dataclass` to `NamedTuple`, which eliminates the overhead of dataclass creation. The line profiler shows the class definition time dropped from 25.5% to 10.3% of total execution time.

**3. Optimized property filtering:**
Replaced the loop-based property deletion (`for attr in ignore: if attr in props: del props[attr]`) with a single dictionary comprehension (`{k: v for k, v in props.items() if k not in ignore_set}`). This avoids multiple dictionary mutations and is more efficient for the typical case of 2 ignored attributes.

**Performance impact by test case:**
- Small tool lists (empty/basic cases): 149-183% faster due to reduced dataclass overhead
- Large tool lists with grouping: 40% faster due to the O(n²) → O(n) optimization 
- Edge case with 1000 identical tools: Minimal impact (1.97% slower) as the bottleneck shifts to ToolProxy creation

The optimizations are most effective when there are many tools with similar properties that need grouping, which is the primary use case for this function.

✅ **Correctness verification report:**

| Test                        | Status            |
| --------------------------- | ----------------- |
| ⚙️ Existing Unit Tests | ✅ **13 Passed** |
| 🌀 Generated Regression Tests | ✅ **6 Passed** |
| ⏪ Replay Tests | 🔘 **None Found** |
| 🔎 Concolic Coverage Tests | ✅ **1 Passed** |
|📊 Tests Coverage       | 97.1% |
<details>
<summary>⚙️ Existing Unit Tests and Runtime</summary>

| Test File::Test Function                       | Original ⏱️   | Optimized ⏱️   | Speedup   |
|:-----------------------------------------------|:--------------|:---------------|:----------|
| `unit/bokeh/test_layouts.py::test_group_tools` | 2.85ms        | 2.52ms         | 13.2%✅   |

</details>

<details>
<summary>🌀 Generated Regression Tests and Runtime</summary>

```python
from collections import defaultdict
# The group_tools function to test (adapted to use the mocks above)
from dataclasses import dataclass
from typing import Any, Callable, TypeVar

# imports
import pytest
from bokeh.layouts import group_tools

# Minimal mock implementations to simulate Bokeh's Tool and ToolProxy

class MockTool:
    """A mock Tool class with customizable properties."""
    def __init__(self, **props):
        self._props = props.copy()

    def properties_with_values(self, *, include_defaults: bool = True, include_undefined: bool = False):
        # Return a copy to avoid accidental mutation in tests
        return self._props.copy()

    def __eq__(self, other):
        return isinstance(other, MockTool) and self._props == other._props

    def __repr__(self):
        return f"MockTool({self._props})"

class MockToolProxy:
    """A mock ToolProxy class that groups tools."""
    def __init__(self, tools):
        self.tools = list(tools)

    def __eq__(self, other):
        return isinstance(other, MockToolProxy) and self.tools == other.tools

    def __repr__(self):
        return f"MockToolProxy({self.tools})"

MergeFn = Callable[[type, list], Any]
from bokeh.layouts import group_tools

# ------------------ UNIT TESTS ------------------

# 1. Basic Test Cases

def test_empty_tools_list():
    # Should return empty list when no tools are provided
    codeflash_output = group_tools([]) # 258μs -> 98.7μs (162% faster)








def test_merge_function_groups_tools():
    # If merge function returns a value, it should be used instead of ToolProxy
    merged_called = []
    def merge(cls, tools):
        merged_called.append((cls, tools))
        return "MERGED"
    t1 = MockTool(foo=1)
    t2 = MockTool(foo=1)
    codeflash_output = group_tools([t1, t2], merge=merge); result = codeflash_output # 262μs -> 105μs (149% faster)











def test_large_number_of_tools_with_merge():
    # Merge function should be called once per group
    calls = []
    def merge(cls, tools):
        calls.append((cls, len(tools)))
        return f"merged_{len(tools)}"
    tools = [MockTool(foo=1) for _ in range(200)] + [MockTool(foo=2) for _ in range(200)]
    codeflash_output = group_tools(tools, merge=merge); result = codeflash_output # 473μs -> 337μs (40.0% faster)


#------------------------------------------------
from collections import defaultdict
# The group_tools function (as provided)
from typing import Any, Callable, TypeVar

# imports
import pytest
from bokeh.layouts import group_tools


# Minimal mock classes to simulate Bokeh's Tool and ToolProxy
class Tool:
    """Mock Tool class with customizable properties."""
    def __init__(self, name=None, **props):
        self.name = name
        self._props = props.copy()
        if name is not None:
            self._props['name'] = name

    def properties_with_values(self, *, include_defaults: bool = True, include_undefined: bool = False):
        return self._props.copy()

    def __repr__(self):
        return f"Tool(name={self.name}, props={self._props})"

class ToolProxy:
    """Mock ToolProxy class that groups multiple tools."""
    def __init__(self, tools):
        self.tools = list(tools)

    def __repr__(self):
        return f"ToolProxy(tools={self.tools})"
MergeFn = Callable[[type, list], Any]  # Simplified for testing
from bokeh.layouts import group_tools

# ------------------- UNIT TESTS BELOW -------------------

# 1. BASIC TEST CASES

def test_empty_tools_list_returns_empty():
    # Should return empty list if input is empty
    codeflash_output = group_tools([]) # 254μs -> 91.6μs (177% faster)







def test_merge_fn_groups_tools():
    # merge function can group tools in a custom way
    t1 = Tool(name="A", foo=1)
    t2 = Tool(name="A", foo=1)
    def merge_fn(cls, tools):
        return ToolProxy(tools)
    codeflash_output = group_tools([t1, t2], merge=merge_fn); result = codeflash_output # 284μs -> 100μs (183% faster)















def test_large_number_of_tools_with_merge_fn():
    # merge function should be called once per class
    class CountingMerge:
        def __init__(self):
            self.count = 0
        def __call__(self, cls, tools):
            self.count += 1
            return ToolProxy(tools)
    merge = CountingMerge()
    tools = [Tool(name="C", foo=42) for _ in range(1000)]
    codeflash_output = group_tools(tools, merge=merge); result = codeflash_output # 739μs -> 754μs (1.97% slower)


#------------------------------------------------
from bokeh.layouts import group_tools

def test_group_tools():
    group_tools([], merge=None, ignore=None)
```

</details>

<details>
<summary>🔎 Concolic Coverage Tests and Runtime</summary>

| Test File::Test Function                                                              | Original ⏱️   | Optimized ⏱️   | Speedup   |
|:--------------------------------------------------------------------------------------|:--------------|:---------------|:----------|
| `codeflash_concolic_cthbg6_3/tmprnifo4ws/test_concolic_coverage.py::test_group_tools` | 290μs         | 98.2μs         | 196%✅    |

</details>

To edit these changes `git checkout codeflash/optimize-group_tools-mh9u58ju` and push.

[Codeflash](https://codeflash.ai)

- [x] issues: fixes #14748
- [x] tests added / passed

### CODE REVIEW

**Comment by bryevdv** [comment]
I'll leave this one to @mattpap or @hoxbro to evaluate.

File: src/bokeh/layouts.py [hunk 1]
```diff
@@ -620,10 +620,9 @@ class ToolEntry:
             computed.append(tool)
         else:
             props = tool.properties_with_values()
-            for attr in ignore:
-                if attr in props:
-                    del props[attr]
-            by_type[tool.__class__].append(ToolEntry(tool, props))
+            # Faster ignore/removal by dictionary comprehension
```
Comment (hoxbro) [inline src/bokeh/layouts.py]: This is AI noise, please remove it.

File: src/bokeh/layouts.py [hunk 2]
```diff
@@ -604,8 +605,7 @@ def traverse(item: LayoutDOM, top_level: bool = False):
 def group_tools(tools: list[Tool | ToolProxy], *, merge: MergeFn[Tool] | None = None,
         ignore: set[str] | None = None) -> list[Tool | ToolProxy]:
     """ Group common tools into tool proxies. """
-    @dataclass
-    class ToolEntry:
+    class ToolEntry(NamedTuple):
```
Comment (hoxbro) [inline src/bokeh/layouts.py]: This shows the opposite for me. 

``` python
import sys
from dataclasses import dataclass
from typing import Any, NamedTuple

Tool = type("Tool", (), {})

@dataclass
class ToolEntry1:
    tool: Tool
    props: Any

class ToolEntry2(NamedTuple):
    tool: Tool
    props: Any

tool, props = Tool(), {}
%timeit ToolEntry1(tool, props)
%timeit ToolEntry2(tool, props)
print(sys.version_info)
```


image

**Review by hoxbro — CHANGES_REQUESTED** [review]

File: src/bokeh/layouts.py (hunk 2 above)
Comment (mattpap) [reply src/bokeh/layouts.py]: Also, the start to any optimizations would be to move the dataclass out of the inner scope. That would most likely dwarf any other improvements.

**Comment by hoxbro** [comment]
Can you update the report based on the latest change?

**Comment by misrasaurabh1** [comment]
Hi @hoxbro , sorry I haven't built the functionality into codeflash.ai to update the performance/correctness report with new changes. Even though i want to build it, i don't see this being ready in the next week.

I think the performance gain might just be better, especially for the very small inputs, with no loss in correctness.

It should be ready to merge from my side.

**Comment by hoxbro** [comment]
> sorry I haven't built the functionality into codeflash.ai

I would greatly appreciate it if you could then perform a manual update, at a minimum, at least update the numbers to reflect the current state.

**Comment by bryevdv** [comment]
@mattpap @hoxbro I don't want to let these linger forever. Do we want to merge it as-is? If not I'd suggest closing.

**Comment by misrasaurabh1** [comment]
im asking @KRRT7 to take it up from my end

**Comment by KRRT7** [comment]
Created associated issue: #14748

The review feedback has been addressed:
- Removed AI noise comment
- Removed unnecessary `ignore_set` variable  
- Kept dataclass (faster than NamedTuple per @hoxbro's benchmark)
- Moved dataclass outside the function (per @mattpap's suggestion)

**Comment by hoxbro** [comment]
Hi @KRRT7, can you update the report (in a new comment, as you did not create the issue) based on the latest change?

**Comment by KRRT7** [comment]
@hoxbro do you mean the performance report?

**Comment by hoxbro** [comment]
yes

**Comment by mosc9575** [comment]
@KRRT7 Do you plan to come back to add the performance report? It passed some time without progress since the request and it would be nice get it closed sooner than later.

**Comment by KRRT7** [comment]
@mosc9575 apologies, will work on this on monday

File: src/bokeh/layouts.py [hunk 3]
```diff
@@ -632,14 +630,16 @@ class ToolEntry:
...
-            for item in list(tail):
-                if item.props == head.props:
-                    group.append(item.tool)
-                    entries.remove(item)
-            entries.remove(head)
+            used[i] = True
+            for j in range(i + 1, len(entries)):
+                if not used[j] and entries[j].props == head.props:
+                    group.append(entries[j].tool)
+                    used[j] = True
```
Comment (mattpap) [inline src/bokeh/layouts.py]: I find this variant more appealing and equally performant:
```py
        items: list[ToolEntry | None] = list(entries)
        for i, head in enumerate(items):
            if head is None:
                continue
            items[i] = None
            group: list[Tool] = [head.tool]
            for j in range(i + 1, len(items)):
                item = items[j]
                if item is not None and item.props == head.props:
                    group.append(item.tool)
                    items[j] = None
```

**Review by mattpap — APPROVED** [review]
Examples failing is unrelated (see PR #14917).
